[Refactoring] Preserve comments in async transform

hamishknight opened the pull request:
Previously we would drop comments between nodes in a BraceStmt, as we printed each node out individually. To remedy this, always make sure we scan backwards to find any preceding comments attached to a node. In addition, make sure we keep track of any SourceLocs which we don't print, but may have comments attached which we want to preserve.

rdar://77401810

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

hamishknight commented:
@swift-ci please test Linux

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4545,32 +4545,135 @@ struct CallbackCondition {
+  }
+
+  /// Add the nodes and comment locs from another NodesToPrint.
+  void addNodes(NodesToPrint OtherNodes) {
```
I'd probably prefer passing in a `const NodesToPrint &` here since we're not avoiding the copy anyway (ditto for the other similar methods).

hamishknight replied:
Talked with Ben offline, going to leave this as-is for now

bnbarham reviewed (approved):
This is awesome, thanks Hamish! Just the one small comment, there'll also be some merge conflicts with mine (I also added a SourceFile)

hamishknight committed:
[Refactoring] Store SourceFile on AsyncConverter

hamishknight committed:
[Refactoring] Preserve comments in async transform

Previously we would drop comments between nodes in
a BraceStmt, as we printed each node out individually.
To remedy this, always make sure we scan backwards
to find any preceding comments attached to a node,
and also keep track of any SourceLocs which we
don't print, but may have comments attached which
we want to preserve.

rdar://77401810

hamishknight commented:
Rebased on top of Ben's change and also added an additional test case to make sure we pickup a trailing comment in a function body if we're converting the function, but not if we're just converting the call.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

hamishknight commented:
@swift-ci please smoke test macOS
